pyrelational/pipeline/generic_pipeline.py

```diff
@@ -0,0 +1,312 @@
+
+logger = logging.getLogger()
+
+
+class GenericPipeline(ABC):
+    """The pipeline facilitates the communication between
+    - DataManager
+    - Model,
+    - ALStrategy,
+    - Oracle (Optional)
+
+    To enact a generic active learning cycle.
+
+    :param data_manager: an pyrelational data manager
+            which keeps track of what has been labelled and creates data loaders for
+            active learning
+    :param model: A pyrelational model
+            which serves as the machine learning model for the data in the
+            data manager
+    :param strategy: A pyrelational active learning strategy
+            implements the informativeness measure and the selection algorithm being used
+    :param oracle: An oracle instance
+            interfaces with various concrete oracles to obtain labels for observations
+            suggested by the strategy
+    """
+
+    def __init__(
+        self,
+        data_manager: GenericDataManager,
+        model: GenericModel,
+        strategy: GenericActiveLearningStrategy,
+        oracle: Optional[GenericOracle] = None,
+    ):
+        super(GenericPipeline, self).__init__()
+        self.data_manager = data_manager
+        self.model = model
+        self.strategy = strategy
+        self.oracle = oracle
+
+        # Mediator pattern for strategy
+        self.strategy._pipeline = self
+
+        # Pipeline meta properties
+        self.iteration = 0
+
+        # Data structures for logging values of interest
+        self.performances = defaultdict(dict)
+        self.labelled_by = {}
+        self.log_labelled_by(data_manager.l_indices, tag="Initialisation")
+
+    def theoretical_performance(self, test_loader: Optional[DataLoader] = None) -> Dict:
+        """Returns the performance of the full labelled dataset against the
+        test data. Typically used for evaluation to establish theoretical benchmark
+        of model performance given all available training data is labelled. The
+        "horizontal" line in area under learning curve plots for active learning
+
+        Would not make much sense when we are doing active learning for the real
+        situation, hence not part of __init__
+
+        :param test_loader: Pytorch Data Loader with
+                test data compatible with model, optional as often the test loader can be
+                generated from data_manager but is here for case when it hasn't been defined
+                or there is a new test set.
+
+        :return: performances
+        """
+        self.model.train(self.train_loader, self.valid_loader)
+
+        # use test loader in data_manager if there is one
+        result = self.model.test(self.test_loader if test_loader is None else test_loader)
+        result = self.compute_hit_ratio(result)
+        self.performances["full"] = result
+
+        # make sure that theoretical best model is not stored
+        self.model.current_model = None
```

Comment:
if self.model.current_model = None has been removed in current_performance, does it also need to be removed here?

Reply:
actually, I guess this serves a different purpose in this case, so here it is necessary?

Reply:
Yes, we should be careful about changing these things without proper checks and it does fall outside of the scope of this PR.

We should however go through repo and make sure that we don't have these kind of attribute setting calls outside of a class. e.g. this should be replaced by a reset_current_model() method in the ModelManager.

Reply:
It is necessary as we don't want to inadvertedly use the model trained on the full training data during any stage of the active learning cycle and give overly optimistic results.

It does raise another question about the situation if someone calls the theoretical_performance at some stage (though I don't see why), it will wipe the currently saved model in the pipeline.

Reply:
okay, moved what Thomas mentioned into a new issue